core/src/main/scala/Elo.scala

```diff
@@ -14,27 +14,35 @@ object KFactor extends OpaqueInt[KFactor]:
  * */
 object Elo extends OpaqueInt[Elo]:
 
-  final class Player(val rating: Elo, val kFactor: KFactor)
-  final class Game(val points: Outcome.Points, val opponentRating: Elo)
-
-  /* 8.3.1
-   * For each game played against a rated player, determine the difference in rating between the player and their opponent.
-   * A difference in rating of more than 400 points shall be counted for rating purposes as though it were a difference of 400 points.  In any tournament, a player may benefit from only one upgrade under this rule, for the game in which the rating difference is greatest. */
-  def playersRatingDiff(a: Elo, b: Elo): Int =
-    Math.min(400, Math.max(-400, b - a))
+  def computeRatingDiff(player: Player, games: Seq[Game]): Int =
+    computeNewRating(player, games) - player.rating
 
-  // https://en.wikipedia.org/wiki/Elo_rating_system#Mathematical_details
   def computeNewRating(player: Player, games: Seq[Game]): Elo =
     val expectedScore = games.foldMap: game =>
       val prd = playersRatingDiff(player.rating, game.opponentRating)
-      1 / (1 + Math.pow(10, prd / 400d))
+      getExpectedScore(prd)
     val achievedScore = games.foldMap(_.points.value)
     val ratingDiff =
-      Math.round(player.kFactor * (achievedScore - expectedScore)).toInt
+      Math.round(player.kFactor * (achievedScore - expectedScore.doubleValue)).toInt
     player.rating + ratingDiff
 
-  def computeRatingDiff(player: Player, games: Seq[Game]): Int =
-    computeNewRating(player, games) - player.rating
+  /* 8.3.1
+   * For each game played against a rated player, determine the difference in rating between the player and their opponent.
+   * A difference in rating of more than 400 points shall be counted for rating purposes as though it were a difference of 400 points.  In any tournament, a player may benefit from only one upgrade under this rule, for the game in which the rating difference is greatest. */
+  def playersRatingDiff(a: Elo, b: Elo): Int =
+    Math.min(400, Math.max(-400, b - a))
+
+  def getExpectedScore(ratingDiff: Int): BigDecimal =
+    val absRatingDiff = ratingDiff.abs
+    val expectedScore = conversionTableFIDE
+      .collectFirst {
+        case (range, pd) if range.contains(absRatingDiff) => pd
+      }
+      .getOrElse {
+        throw new IllegalArgumentException(s"Invalid rating difference: $ratingDiff")
```

Comment:
we pretty much never throw exceptions in scala. Return `Option[BigDecimal]` if the function is not defined for all its possible, inputs or default to a given value. Here defaulting to `0.5` is correct.